### ISSUE TITLE
react/require-default-props

### ISSUE THREAD

**Issue by xareelee** [yannickcr/eslint-plugin-react#1212]
I meet a bug with `react/require-default-props`:

```js
class ViewContainer extends React.Component {

  props: {
    getMsgCodeErrorMsg?: string,
  };

  ...
}
```

The should not raise an error for `react/require-default-props` with the linter.

The prop `getMsgCodeErrorMsg` is optional with a exact `string` type, which means that you can provide no default value at the first time, but if you pass a value to the prop `getMsgCodeErrorMsg`, it needs to be a non-null string value.

Am I right?

**Comment by ljharb** [comment]
You're right that this is a bug in Flow (you should be able to specify a defaultProp of `undefined`, or possibly `null`).

This is a duplicate of #1210, and the Flow issue is https://github.com/facebook/flow/issues/3982.

**Comment by xareelee** [comment]
Any other case for `react/require-default-props` warning:

```js
type FooProps = {
  a: string,
  b?: boolean,  // optional
}

const Foo = ({
  a,
  b = false,  // providing a default value for props
}: FooProps) => {
  // ...
}
```

This also reports a lint error `react/require-default-props` in this stateless React component form:

```
14:3  error  propType "b" is not required, but has no corresponding defaultProp declaration          react/require-default-props
```

**Comment by ljharb** [comment]
That's intentional. SFC default values are not a substitute for defaultProps.

**Comment by xareelee** [comment]
I don't get why it can't. Any reason?

```js
type FooProps = {
  ...
}

const Foo = ({
  ...
}: FooProps) => {
  // long implementation
  ...
}

Foo.defaultProps = {
  ...
}
```

If the implementation of the component `Foo` is too long, the type definition and the default value definition would be separated so far. It's low cohesiveness.

**Comment by ljharb** [comment]
Then the solution is to make the implementation shorter, clearly :-)

For one, React applies defaultProps on its own; it can't read into default values.
